### DISCUSSION TITLE
App with slightly different functionality by DNS name

### DISCUSSION THREAD

**Discussion by gettalong** [jeremyevans/roda#316] [General]
Beforehand: Thanks for Roda!

I'm a first-time Roda user, read the "Mastering Roda" manual and looked through Roda code and API documentation. So everything is still quite new for me :smiley: 

I'm porting a rather small Rails app (just a service web app without even a database) to Roda and I want to ask for advice with regards to structuring the routes. The app provides different functionality based on the hostname, think an admin part and a user part.

After reading the documentation I see that I can use the "header_matchers" plugin to route by hostname:

~~~ ruby
class App < Roda
  plugin :header_matchers

  route do |r|
    r.on(host: 'user.example.com') do
      # user staff
    end

    # Do we even need to use r.on here?
    r.on(host: 'admin.example.com') do
      # admin stuff
    end
end
~~~

Both, the user and admin parts share common functionality, like log-in. So I think that such functionality would need to be routed before routing by host name - is this correct?

And both parts are sufficiently complex/have 10+ routes that putting the routes into separate files would be advantageous. I thought about using the 'hash_branches' plugin but I'm not sure how to integrate that with the host matching.

Any advice on best practices in such a case would be really helpful - thank you!

**Comment by jeremyevans** [comment]
If you want to have completely different routes available per host, that seems to be a reason to use the `header_matchers` plugin in a manner similar to your example.  However, if you just want different functionality per subdomain, while having routing mostly the same, you probably don't want to use that.  A simple `case` statement may work, something like:

```ruby
  route do |r|
    case env['HTTP_HOST']
    when 'user.example.com'
      # user staff
    when 'admin.example.com'
      # admin stuff
    end

    # routing code
  end
```

You could definitely move the `case` statement to a separate method if you wanted to use it in multiple places.  Hopefully this helps.  If you have more questions, please let me know.

**Reply by gettalong** [reply]
Thank you, that helps indeed! I still need to wrap my head around the different control flow of Roda's routing and remember that it's just method invocations.
